rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/package.scala

```diff
@@ -37,11 +37,19 @@ package object matcher {
   def _freeMap[F[_]](implicit ev: _freeMap[F]): _freeMap[F] = ev
   def _short[F[_]](implicit ev: _short[F]): _short[F]       = ev
 
-  private[matcher] def runFirstWithCost[F[_]: Monad, A](
-      f: MatcherMonadT[F, A],
-      initCost: Cost
+  private[matcher] def run[F[_]: Monad: _cost, A](
+      f: MatcherMonadT[F, A]
+  ): F[Stream[(FreeMap, A)]] =
+    StreamT.run(f.run(emptyMap))
+
+  private[matcher] def runFirstWithCost[F[_]: Monad: _cost, A](
+      f: MatcherMonadT[F, A]
   ): F[(Cost, Option[(FreeMap, A)])] =
-    StreamT.run(StreamT.dropTail(f.run(emptyMap))).map(_.headOption).run(initCost)
+    for {
+      cost <- _cost[F].get
+      s <- StreamT
+            .run(StreamT.dropTail(f.run(emptyMap)))
+    } yield ((cost, s.headOption))
```

Comment:
Isn't it that you get the cost prior to running the matcher monad? Wouldn't this mean that the cost you get doesn't account for performing the match?